### ISSUE TITLE
Refitting quantile regression model does not work when target is large

### ISSUE THREAD

**Issue by takeyama0** [microsoft/LightGBM#5760]
## Description
Refitting quantile regression model does not work when target is large.
In particular, if the target is sufficiently larger than the gradient of the pinball loss function, typically the target y >> 1, the quantile regression refit will not work.

I think this is due to the fact that the algorithm for determining the output of each leaf is different between the normal quantile regression model construction and refitting. 
During normal training, the output of a leaf is the quantile of the data set contained in each leaf, rather than gradient boosting as described [here](https://github.com/microsoft/LightGBM/blob/master/src/objective/regression_objective.hpp#L524). On the other hand, in refit, the leaf values are updated by gradient boosting maybe [here](https://github.com/microsoft/LightGBM/blob/master/src/treelearner/serial_tree_learner.cpp#L212).

Because the gradient of the pinball loss is a constant between -1~1 independent of target size, convergence is slow when the target is sufficiently larger than 1.

## Reproducible example
``` python
import numpy as np
import lightgbm as lgb
from matplotlib import pyplot as plt

# setup for input data
X_shape = (10000, 1)
y_scale = 1 # 0.01

# create input data
X = np.random.random(X_shape)
y_true = np.sin(10 * X[:,0])
y = y_true +  0.5 * np.random.normal(0, 1, X_shape[0])

# train 90% quantile model
q_90_params = {
    "objective": "quantile",
    "alpha": 0.9,
}
lgbm_q_90 = lgb.train(params=q_90_params, train_set=lgb.Dataset(X, y))

# train 50% quantile model
q_50_params = {
    "objective": "quantile",
    "alpha": 0.5,
}
lgbm_q_50 = lgb.train(params=q_50_params, train_set=lgb.Dataset(X, y))

# train 90% quantile model by refitting 50% quantile model
lgbm_q_90_refitted = lgb.train(params=q_50_params, train_set=lgb.Dataset(X, y))
lgbm_q_90_refitted.params['alpha']=0.9
lgbm_q_90_refitted.refit(X, y, decay_rate=0.)

# plot output
fig, ax = plt.subplots()
ax.scatter(X[:,0], y, label='observed y', s=1)
ax.scatter(X[:,0], lgbm_q_90.predict(X), label='90% quantile', color="orange", s=1)
ax.scatter(X[:,0], lgbm_q_50.predict(X), label='50% quantile', color="black", s=1)
ax.scatter(X[:,0], lgbm_q_90_refitted.predict(X), label='refitted 90% quantile', color="red", s=1)
plt.legend(markerscale=5)
plt.show()
```
![quantile_reft](https://user-images.githubusercontent.com/78151275/222419698-225272e4-d064-4644-b503-2c928dc0013f.png)

If we scale the target value by 0.01, the refit works fine like below. (modify y_scale to 0.01 at the beginning of the code above )

![quantile_reft_scale_001](https://user-images.githubusercontent.com/78151275/222420047-d695d4d0-fd0d-48e3-9717-2b3dc7a9ff10.png)

**Comment by jameslamb** [comment]
Thank you for the excellent write-up, and sorry it's taken so long to get you a response!

@jmoralez is this something you could look into?
